File: src/vcmp/functions/marker.py
from vcmp.__export import funcs
from vcmp.vec import Vector
from vcmp.instance import RGB, Marker

def create_marker(
    model: int,
    world: int,
    pos: Vector,
    scale: int,
    color: RGB
):
    """

    Create a marker

    :param model: The model of the object.
    :param world: The world to create the object in.
    :param pos: The position of the object.
    :param alpha: The alpha of the object.
    :return: The object created.
    """
    return Marker(
        funcs.create_coord_blip(-1, world, pos.x, pos.y, pos.z, scale, color.to_alpha(), model)
    )

File: crates/vcmp_bindings/src/func/marker.rs
use crate::func::VcmpFunctions;
use crate::utils::{Color, Marker, Vector};

pub trait MarkerMethods {
    fn create_marker(
        &self,
        world: i32,
        pos: (f32, f32, f32),
        scale: i32,
        color: Color,
        sprite: i32,
        index: Option<i32>,
    ) -> i32;
    fn destory_marker(&self, marker: i32);
    fn get_marker_info(&self, marker: i32) -> Marker;
}

impl MarkerMethods for VcmpFunctions {
    fn create_marker(
        &self,
        world: i32,
        pos: (f32, f32, f32),
        scale: i32,
        color: Color,
        sprite: i32,
        index: Option<i32>,
    ) -> i32 {
        let idx = index.unwrap_or(-1);
        (self.inner.CreateCoordBlip)(idx, world, pos.0, pos.1, pos.2, scale, color.into(), sprite)
    }
    fn destory_marker(&self, marker: i32) {
        (self.inner.DestroyCoordBlip)(marker);
    }

    fn get_marker_info(&self, marker: i32) -> Marker {
        let (mut world, mut x, mut y, mut z, mut scale, mut color, mut sprite) =
            (0, 0.0, 0.0, 0.0, 0, 0, 0);
        (self.inner.GetCoordBlipInfo)(
            marker,
            &mut world,
            &mut x,
            &mut y,
            &mut z,
            &mut scale,
            &mut color,
            &mut sprite,
        );

        Marker {
            marker,
            world,
            position: Vector { x, y, z },
            scale,
            color: Color::from_argba,
            sprite,
        }
    }
}
